### PR TITLE
Update and Upsert functionality for PostgreSQL plugins. (CloudSQL PostgreSQL, Aurora PostgreSQL)

### PR DESCRIPTION
**Pull request by prince-cs**

### CODE REVIEW

File: database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java [hunk 1]
```diff
@@ -212,11 +244,36 @@ public void write(DataOutput out) throws IOException {
...
       Schema.Field field = record.getSchema().getField(columnType.getName());
       writeToDB(stmt, field, i);
+      if (fillUpdateParams(updatedKeyList, columnType)) {
+        modifiableColumnTypes.add(columnType);
+      }
     }
+    if (operationName != null && relationTableKey != null) {
+      // index to store the latest place of question mark to be allocated
+      int j = columnTypes.size();
+      for (int i = j; i < modifiableColumnTypes.size(); i++) {
```
Comment (dhiraj0116) [inline database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java]: Avoid using variable names as 'i' and 'j'

File: database-commons/src/main/java/io/cdap/plugin/db/config/DatabaseSinkConfig.java [hunk 2]
```diff
@@ -72,4 +72,14 @@ public interface DatabaseSinkConfig extends DatabaseConnectionConfig {
   default void validate(FailureCollector collector) {
     // no-op
   }
+
+  /**
+   * @return the operation to be performed on the query
+   */
+  String getOperationName();
```
Comment (dhiraj0116) [inline database-commons/src/main/java/io/cdap/plugin/db/config/DatabaseSinkConfig.java]: Any specific reason of using 'default' access modifier here?

File: database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java [hunk 3]
```diff
@@ -212,11 +244,36 @@ public void write(DataOutput out) throws IOException {
...
+        ColumnType columnType = modifiableColumnTypes.get(i);
+        Schema.Field field = record.getSchema().getField(columnType.getName());
+        writeToDB(stmt, field, i);
+      }
+    }
+  }
+
+  private boolean fillUpdateParams(List<String> updatedKeyList, ColumnType columnType) {
+    if (operationName != null && operationName.equalsIgnoreCase("UPDATE") &&
```
Comment (dhiraj0116) [inline database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java]: instead of 'update' use already defined constant from ETLDBOutputFormat class.

File: database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java [hunk 4]
```diff
@@ -114,6 +126,38 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
...
   }
+  public void validateOperations(FailureCollector collector, T dbSinkConfig, @Nullable Schema inputSchema) {
+    if (dbSinkConfig.containsMacro(OPERATION_NAME) || dbSinkConfig.containsMacro(RELATION_TABLE_KEY)) {
+      return;
+    }
+    String operation = dbSinkConfig.getOperationName();
+    if (operation.equalsIgnoreCase("INSERT")) {
+      return;
+    }
+    if ((operation.equalsIgnoreCase("UPDATE") || operation.equalsIgnoreCase("UPSERT"))
```
Comment (dhiraj0116) [inline database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java]: Use constants

File: database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java [hunk 5]
```diff
@@ -67,14 +72,27 @@ public RecordWriter<K, V> getRecordWriter(TaskAttemptContext context) throws IOE
     String tableName = dbConf.getOutputTableName();
     String[] fieldNames = dbConf.getOutputFieldNames();
     final int batchSize = conf.getInt(COMMIT_BATCH_SIZE, DEFAULT_COMMIT_BATCH_SIZE);
+    final String operationName = conf.get(OPERATION_NAME);
+    String relationTableKey = null;
```
Comment (dhiraj0116) [inline database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java]: Why default access modifier?

File: database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java [hunk 6]
```diff
@@ -191,6 +209,9 @@ private Connection getConnection(Configuration conf) {
     }
     return connection;
   }
+  public String constructUpsertQuery(String table, String[] fieldNames, String[] listKeys) {
```
Comment (dhiraj0116) [inline database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java:226]: This method always returns 'null'?
what's the purpose here?

Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java:226]: This method is used to create Upsert query and the query for upsert is different for all plugins. This method is acting as a parent method and for postgreSQL plugins, this method is getting overridden for the upsert query creation for postgrSQL plugins. 

As per the requirement only PostgreSQL plugins were required to have upsert query. If in future, any other plugin is asked to have upsert query then the same procedure will be followed.

File: database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java (hunk 1 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java]: done.

File: database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java (hunk 3 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java]: done

File: database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java (hunk 4 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java]: done

File: database-commons/src/main/java/io/cdap/plugin/db/config/DatabaseSinkConfig.java (hunk 2 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/config/DatabaseSinkConfig.java]: The implementation of these two added methods are being used by the parent sink which are used to pass values at runtime configurations. Therefore, using default access modifier.

File: database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java (hunk 5 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java]: access modifier are not allowed here as it is inside a public method which is getRecordWriter.

File: aurora-postgresql-plugin/widgets/AuroraPostgres-batchsink.json [hunk 7]
```diff
@@ -95,6 +124,20 @@
     }
   ],
   "outputs": [],
+  "filters": [
```
Comment (vikasrathee-cs) [inline aurora-postgresql-plugin/widgets/AuroraPostgres-batchsink.json]: remove this filter condition and let relationTableKey field always visible. Put a validation if user has selected upsert or update, then this field can not be null.
The reason for removing this filter is if user selects operation name as macro, this field will not be visible to set it as macro.

File: database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java [hunk 8]
```diff
@@ -65,6 +69,22 @@ public class DBRecord implements Writable, DBWritable, Configurable {
    */
   protected List<ColumnType> columnTypes;
 
+  /**
+   * Need to cache column types to set fields of input record for where clause on {@link PreparedStatement}
+   * in the right order.
+   */
+  protected List<ColumnType> modifiableColumnTypes;
```
Comment (vikasrathee-cs) [inline database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java:75]: what is the use of this field?

File: database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java [hunk 9]
```diff
@@ -171,17 +222,24 @@ public void prepareRun(BatchSinkContext context) {
     }
 
     // Get Hadoop configuration object
-    Configuration configuration = configAccessor.getConfiguration();
+//    Configuration configuration = configAccessor.getConfiguration();
```
Comment (vikasrathee-cs) [inline database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java]: why this is commented?

File: database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java [hunk 10]
```diff
@@ -53,6 +53,12 @@
   // Batch size before submitting a batch to the SQL engine. If set to 0, no batches will be submitted until commit.
   public static final String COMMIT_BATCH_SIZE = "io.cdap.plugin.db.output.commit.batch.size";
   public static final int DEFAULT_COMMIT_BATCH_SIZE = 1000;
+  public static final String OPERATION_NAME = "io.cdap.plugin.db.operation.name";
+  public static final String RELATION_TABLE_KEY = "io.cdap.plugin.db.relation.table.key";
+
+  public static final String UPDATE = "update";
```
Comment (vikasrathee-cs) [inline database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java]: create an enum for operation

File: database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java (hunk 8 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java:75]: In case of update, the no. of question marks which are to be written in the prepared statement throw Out of bounds exception as column types contains all the columns of the table. So in order to remove that exception, a new list is being created namely modifiableColumnTypes, which is getting updated to contain all the information about the column names on which the updation needs to take place.

File: database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java (hunk 9 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java]: The configuration are being set in another way and it was not required anymore. Therefore, commented out the code.

File: database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java (hunk 6 above)
Comment (vikasrathee-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java:226]: this method should return this.constructUpsertQuery, so that child class method can be invoked, if this has child class reference

File: database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java [hunk 11]
```diff
@@ -208,4 +230,45 @@ public String constructQuery(String table, String[] fieldNames) {
...
+  public String constructQueryUpdateAndUpsert(String tableName, String[] fieldNames, String operationName,
+                                              String[] listKeys) {
+    String query = null;
+    if (operationName.equalsIgnoreCase(UPDATE)) {
+      query = constructUpdateQuery(tableName, fieldNames, listKeys);
+    } else if (operationName.equalsIgnoreCase(UPSERT)) {
+      query = constructUpsertQuery(tableName, fieldNames, listKeys);
+    }
+    return query;
+  }
```
Comment (vikasrathee-cs) [inline database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java:267]: Add a space. if these methods are not getting invoked outside this class, make it private or protected. Also write comments on top of it.

File: postgresql-plugin/src/test/java/io/cdap/plugin/postgres/PostgresSinkTestRun.java [hunk 12]
```diff
@@ -53,6 +59,8 @@
  * Test for ETL using databases.
  */
 public class PostgresSinkTestRun extends PostgresPluginTestBase {
+  private static final String CLASS_NAME = "org.postgresql.Driver";
+  private static final String URL = "jdbc:postgresql://localhost:5432/demodb";
```
Comment (vikasrathee-cs) [inline postgresql-plugin/src/test/java/io/cdap/plugin/postgres/PostgresSinkTestRun.java]: why this is added, we are suppose to write junits only

File: postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresDBRecord.java [hunk 13]
```diff
@@ -38,6 +38,10 @@ public class PostgresDBRecord extends DBRecord {
   public PostgresDBRecord(StructuredRecord record, List<ColumnType> columnTypes) {
```
Comment (vikasrathee-cs) [inline postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresDBRecord.java]: remove unused constructor

File: postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresDBRecord.java [hunk 14]
```diff
@@ -85,7 +89,7 @@ private Object createPGobject(String type, String value, ClassLoader classLoader
   @Override
   protected void writeToDB(PreparedStatement stmt, Schema.Field field, int fieldIndex) throws SQLException {
     int sqlIndex = fieldIndex + 1;
-    ColumnType columnType = columnTypes.get(fieldIndex);
+    ColumnType columnType = modifiableColumnTypes.get(fieldIndex);
     if (PostgresSchemaReader.STRING_MAPPED_POSTGRES_TYPES_NAMES.contains(columnType.getTypeName()) ||
       PostgresSchemaReader.STRING_MAPPED_POSTGRES_TYPES.contains(columnType.getType())) {
       stmt.setObject(sqlIndex, createPGobject(columnType.getTypeName(),
```
Comment (vikasrathee-cs) [inline postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresDBRecord.java:94]: what is the use of createPGobject and can it affect our implementation if any of the column falls under this category.

Comment (prince-cs) [reply postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresDBRecord.java:94]: It will not have any impact on our implementation.

File: database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java (hunk 11 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java:267]: this method is being used outside of this class, therefore public.

File: database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java (hunk 10 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java]: done

File: database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java (hunk 6 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java:226]: done

File: aurora-postgresql-plugin/widgets/AuroraPostgres-batchsink.json (hunk 7 above)
Comment (prince-cs) [reply aurora-postgresql-plugin/widgets/AuroraPostgres-batchsink.json]: done

File: database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java (hunk 11 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java:267]: done

File: postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresDBRecord.java (hunk 13 above)
Comment (prince-cs) [reply postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresDBRecord.java]: done

File: postgresql-plugin/src/test/java/io/cdap/plugin/postgres/PostgresSinkTestRun.java (hunk 12 above)
Comment (prince-cs) [reply postgresql-plugin/src/test/java/io/cdap/plugin/postgres/PostgresSinkTestRun.java]: Working on making the changes accordingly.

Comment (prince-cs) [reply postgresql-plugin/src/test/java/io/cdap/plugin/postgres/PostgresSinkTestRun.java]: resolved

File: database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java (hunk 9 above)
Comment (vikasrathee-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java]: remove it from comments

Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java]: resolved

File: database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java (hunk 10 above)
Comment (vikasrathee-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java]: getOperation should return enum value instead of string. e.g.   public Operation getOperation() {
    return Strings.isNullOrEmpty(operation) ? Operation.INSERT : Operation.valueOf(operation.toUpperCase());
  }

File: postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresDBRecord.java (hunk 14 above)
Comment (vikasrathee-cs) [reply postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresDBRecord.java:94]: have you tested with such type of column in postgres, if not please do it

File: database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java [hunk 15]
```diff
@@ -114,6 +127,45 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
       validateSchema(collector, driverClass, dbSinkConfig.getTableName(), inputSchema, dbSinkConfig.getDBSchemaName());
     }
   }
+  public void validateOperations(FailureCollector collector, T dbSinkConfig, @Nullable Schema inputSchema) {
+    if (dbSinkConfig.containsMacro(OPERATION_NAME) || dbSinkConfig.containsMacro(RELATION_TABLE_KEY)) {
+      return;
+    }
+    String operation = dbSinkConfig.getOperationName();
+    String relationTableKey = dbSinkConfig.getRelationTableKey();
+    if (operation.equalsIgnoreCase(Operation.INSERT.getName()) && relationTableKey == null) {
```
Comment (vikasrathee-cs) [inline database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java]: this condition check is not required, if user puts some wrong value also, we will not consider it

File: database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java [hunk 16]
```diff
@@ -114,6 +127,45 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
...
+      return;
+    }
+    if (operation.equalsIgnoreCase(Operation.INSERT.getName()) && relationTableKey != null) {
```
Comment (vikasrathee-cs) [inline database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java]: remove this also

File: database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java [hunk 17]
```diff
@@ -114,6 +127,45 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
       validateSchema(collector, driverClass, dbSinkConfig.getTableName(), inputSchema, dbSinkConfig.getDBSchemaName());
     }
   }
+  public void validateOperations(FailureCollector collector, T dbSinkConfig, @Nullable Schema inputSchema) {
+    if (dbSinkConfig.containsMacro(OPERATION_NAME) || dbSinkConfig.containsMacro(RELATION_TABLE_KEY)) {
+      return;
+    }
+    String operation = dbSinkConfig.getOperationName();
```
Comment (vikasrathee-cs) [inline database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java]: this method should return enum value

File: database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java [hunk 18]
```diff
@@ -86,6 +91,13 @@
   protected List<String> columns;
   protected List<ColumnType> columnTypes;
   protected String dbColumns;
+  private ConnectionConfigAccessor configAccessor;
+  private Configuration configuration;
+
+  {
```
Comment (vikasrathee-cs) [inline database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java]: instead of instance block, initialise these in constructor

File: database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java [hunk 19]
```diff
@@ -114,6 +127,45 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
...
+      return;
+    }
+    if ((operation.equalsIgnoreCase(Operation.UPDATE.getName()) ||
+      operation.equalsIgnoreCase(Operation.UPSERT.getName())) && relationTableKey == null) {
+      collector.addFailure(
+          "Table key must be set if the operation is 'Update' or 'Upsert'.", null)
+        .withConfigProperty(RELATION_TABLE_KEY).withConfigProperty(OPERATION_NAME);
+      return;
+    }
+    if (inputSchema == null) {
```
Comment (vikasrathee-cs) [inline database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java:148]: Test a use case where input schema is null due to any connection property as macro , will it work even then. try running the full pipeline

File: database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java [hunk 20]
```diff
@@ -170,18 +225,22 @@ public void prepareRun(BatchSinkContext context) {
...
-    // Get Hadoop configuration object
-    Configuration configuration = configAccessor.getConfiguration();
-
     // Configure batch size if specified in pipeline arguments.
     if (context.getArguments().has(ETLDBOutputFormat.COMMIT_BATCH_SIZE)) {
       configuration.set(ETLDBOutputFormat.COMMIT_BATCH_SIZE,
                         context.getArguments().get(ETLDBOutputFormat.COMMIT_BATCH_SIZE));
     }
 
+    addOutputContext(context);
```
Comment (vikasrathee-cs) [inline database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java:231]: make sure use of 'this' not required here. test it once that it is calling child class

File: database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java [hunk 21]
```diff
@@ -67,14 +70,27 @@ public RecordWriter<K, V> getRecordWriter(TaskAttemptContext context) throws IOE
     String tableName = dbConf.getOutputTableName();
     String[] fieldNames = dbConf.getOutputFieldNames();
     final int batchSize = conf.getInt(COMMIT_BATCH_SIZE, DEFAULT_COMMIT_BATCH_SIZE);
+    final String operationName = conf.get(OPERATION_NAME);
+    String relationTableKey = null;
+    String[] listKeys = new String[0];
```
Comment (vikasrathee-cs) [inline database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java]: initialize it with nulll and null checks should be there in further methods where we are passing it as argument

File: database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java [hunk 22]
```diff
@@ -67,14 +70,27 @@ public RecordWriter<K, V> getRecordWriter(TaskAttemptContext context) throws IOE
...
 
     if (fieldNames == null) {
       fieldNames = new String[dbConf.getOutputFieldCount()];
     }
 
     try {
       Connection connection = getConnection(conf);
-      PreparedStatement statement = connection.prepareStatement(constructQuery(tableName, fieldNames));
+      PreparedStatement statement;
+      if (relationTableKey != null) {
```
Comment (vikasrathee-cs) [inline database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java]: check for listkeys instead, relationTableKey can be used inside the previous if block only and also add a check for operation here

File: database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java [hunk 23]
```diff
@@ -212,11 +244,34 @@ public void write(DataOutput out) throws IOException {
    * @param stmt the {@link PreparedStatement} to write the {@link StructuredRecord} to
    */
   public void write(PreparedStatement stmt) throws SQLException {
+    List<String> updatedKeyList = new ArrayList<>();
+    if (relationTableKey != null) {
```
Comment (vikasrathee-cs) [inline database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java]: check for operation also

File: database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java [hunk 24]
```diff
@@ -212,11 +244,34 @@ public void write(DataOutput out) throws IOException {
...
+    modifiableColumnTypes = new ArrayList<>(columnTypes);
     for (int i = 0; i < columnTypes.size(); i++) {
       ColumnType columnType = columnTypes.get(i);
       Schema.Field field = record.getSchema().getField(columnType.getName());
       writeToDB(stmt, field, i);
+      if (fillUpdateParams(updatedKeyList, columnType)) {
+        modifiableColumnTypes.add(columnType);
+      }
     }
+    if (operationName != null && relationTableKey != null) {
```
Comment (vikasrathee-cs) [inline database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java:293]: what will be the value of operation here, can it be null in case of insert for other plugins?add check for value also.
And writetoDB method can be called twice here. This logic is not looking good, please have a look
Also to make it more readable, add proper comments

File: database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java [hunk 25]
```diff
@@ -212,11 +244,34 @@ public void write(DataOutput out) throws IOException {
...
+      for (int i = columnTypes.size(); i < modifiableColumnTypes.size(); i++) {
+        ColumnType columnType = modifiableColumnTypes.get(i);
+        Schema.Field field = record.getSchema().getField(columnType.getName());
+        writeToDB(stmt, field, i);
+      }
```
Comment (vikasrathee-cs) [inline database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java:298]: how Upsert will be perfomed here?

File: oracle-plugin/src/main/java/io/cdap/plugin/oracle/OracleSourceDBRecord.java [hunk 26]
```diff
@@ -117,7 +117,7 @@ protected void handleField(ResultSet resultSet, StructuredRecord.Builder recordB
 
   @Override
   protected void writeToDB(PreparedStatement stmt, @Nullable Schema.Field field, int fieldIndex) throws SQLException {
-    int sqlType = columnTypes.get(fieldIndex).getType();
+    int sqlType = modifiableColumnTypes.get(fieldIndex).getType();
```
Comment (vikasrathee-cs) [inline oracle-plugin/src/main/java/io/cdap/plugin/oracle/OracleSourceDBRecord.java]: revert this change

File: postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresETLDBOutputFormat.java [hunk 27]
```diff
@@ -0,0 +1,87 @@
...
+ * Class that extends {@link ETLDBOutputFormat} to implement the abstract methods
+ */
+public class PostgresETLDBOutputFormat extends ETLDBOutputFormat {
+
+  /**
+   * This method is used to construct the upsert query for PostgreSQL
+   * @param table - Name of the table
+   * @param fieldNames - All the columns of the table
+   * @param listKeys - The columns to be updated
+   * @return Upsert query in the form of string
```
Comment (vikasrathee-cs) [inline postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresETLDBOutputFormat.java:34]: write in comment a sample upsert query

File: postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresETLDBOutputFormat.java [hunk 28]
```diff
@@ -0,0 +1,87 @@
...
+          query.append(listKeys[i]);
+          if (i != listKeys.length - 1) {
+            query.append(", ");
+          }
+        }
+
+        query.append(")").append(" DO UPDATE SET ");
+
+        for (i = 0; i < listKeys.length; ++i) {
+          query.append(listKeys[i]).append(" = EXCLUDED.").append(listKeys[i]);
```
Comment (vikasrathee-cs) [inline postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresETLDBOutputFormat.java]: when excluded will be used, can it throw error if there is unique constraint on a field that is not part of list keys field

File: postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresSink.java [hunk 29]
```diff
@@ -129,6 +138,10 @@ public static class PostgresSinkConfig extends AbstractDBSpecificSinkConfig {
       "The timeout is specified in seconds and a value of zero means that it is disabled")
     @Nullable
     public Integer connectionTimeout;
+    public PostgresSinkConfig(@Nullable String operationName, @Nullable String relationTableKey) {
```
Comment (vikasrathee-cs) [inline postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresSink.java]: make it default and put visiblefor testing annotation.

**Review by vikasrathee-cs — CHANGES_REQUESTED** [review]
Please have a look at review comments. Add comments on top of complex code to make it understandable.

File: database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java (hunk 20 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java:231]: yes i have checked that and it is working fine.

File: database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java (hunk 25 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java:298]: in case of mysql plugin, this piece of code will be used for upsert as it will have question marks to be filled unlike other db plugins for upsert.

File: database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java (hunk 24 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java:293]: operation name cannot be null as by default for other plugins the value being passed is insert. and it will not be called twice because of the modifiableColumnTypes. This will contain the updated size based on the listKeys added to be filled while writing. writeToDB will only be called when modifiableColumnTypes size will be updated. Basically, if it will have field to be updated then the for loop will be executed.

File: database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java (hunk 23 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java]: done

File: database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java (hunk 15 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java]: done

File: database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java (hunk 16 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java]: done

File: database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java (hunk 17 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java]: done

File: database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java (hunk 18 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java]: done

File: database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java (hunk 21 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java]: done

File: database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java (hunk 22 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java]: done

File: database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java (hunk 25 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java:298]: this code was written to fill the question marks we get in case of update and upsert (only mysql based plugins).

File: postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresETLDBOutputFormat.java (hunk 27 above)
Comment (prince-cs) [reply postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresETLDBOutputFormat.java:34]: done

File: postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresSink.java (hunk 29 above)
Comment (prince-cs) [reply postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresSink.java]: done

File: oracle-plugin/src/main/java/io/cdap/plugin/oracle/OracleSourceDBRecord.java (hunk 26 above)
Comment (prince-cs) [reply oracle-plugin/src/main/java/io/cdap/plugin/oracle/OracleSourceDBRecord.java]: done

File: database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java (hunk 19 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java:148]: done

File: database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java (hunk 10 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/ETLDBOutputFormat.java]: done

File: database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java (hunk 15 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java]: this condition is required as for INSERT operation we don't need to go further and check for field values.

File: database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java (hunk 16 above)
Comment (prince-cs) [reply database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java]: We require this condition as well, because in case of insert the relation keys are supposed to be null.

File: postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresETLDBOutputFormat.java (hunk 28 above)
Comment (prince-cs) [reply postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresETLDBOutputFormat.java]: It throws error and it fails the pipeline.

Comment (prince-cs) [reply postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresETLDBOutputFormat.java]: Checked the scenarios on this as expected and the pipeline fails if no unique constraint is found for ON CONFLICT clause. The exception is thrown by the respected database and is captured in logs.

File: postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresDBRecord.java (hunk 14 above)
Comment (prince-cs) [reply postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresDBRecord.java:94]: tested and working fine.
